### PR TITLE
Aft-Foil Surface Loss Upweighting: 1.5x weight on aft-foil nodes

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

p_tan (28.341) is the hardest metric and the primary bottleneck. It measures surface pressure accuracy on the aft foil in tandem configurations, which is dominated by complex wake interactions from the fore foil.

Currently, the surface loss treats all surface nodes equally — fore-foil and aft-foil nodes contribute the same loss weight. But aft-foil predictions are systematically harder and more important for p_tan.

**Upweighting aft-foil surface loss by 1.5x** directs more optimization pressure toward the harder aft-foil predictions. This is analogous to class imbalance weighting in classification — the "harder class" (aft-foil) gets more attention.

**Why this should help:**
1. **Gradient allocation:** With equal weights, the optimizer spends equal effort on fore-foil (easier, already accurate) and aft-foil (harder, main source of p_tan error). 1.5x shifts gradient budget toward the bottleneck.
2. **Different from OHNM:** Online Hard Node Mining (#2249, in flight) upweights hard NODES within each foil. This upweights the entire AFT FOIL — they're orthogonal and complementary.
3. **Backed by data:** The higher weight decay experiment (#2246) showed p_tan improved with stronger regularization. Aft-foil upweighting is a more targeted way to improve p_tan without the p_in regression from global regularization.

**Risk:** If 1.5x is too aggressive, fore-foil accuracy (p_in) could degrade as the optimizer underweights it. The 1.5x factor is conservative — enough to shift gradient allocation meaningfully without starving fore-foil training.

**Confidence:** Medium. Loss weighting is a direct, well-understood lever. The uncertainty is in the exact multiplier.

### Code changes in `cfd_tandemfoil/train.py`

**Step 1: Add config flag**

In the `Config` class, add:
```python
aft_foil_loss_weight: float = 1.0   # multiplier for aft-foil (boundary_id=7) surface loss (1.0 = no change)
```

**Step 2: Apply the weight in the surface loss computation**

Find where the surface loss is computed. There should be a section that computes per-node losses on surface nodes. The aft-foil surface nodes are identified by `boundary_id == 7` (or a similar mask like `saf_norm` or the foil-2 mask).

Apply the weight BEFORE reduction:
```python
# After computing per-node surface loss but BEFORE mean/sum reduction:
if cfg.aft_foil_loss_weight != 1.0:
    # boundary_id 7 = aft-foil surface nodes
    aft_mask = (boundary_id == 7)  # adjust variable name as needed
    node_weights = torch.ones_like(surface_loss_per_node)
    node_weights[aft_mask] = cfg.aft_foil_loss_weight
    surface_loss_per_node = surface_loss_per_node * node_weights
```

**IMPORTANT:** Only modify the surface loss term, not the volume loss. The weight should apply equally to all channels (Ux, Uy, p) on aft-foil nodes. Look at the existing code to understand how boundary IDs are stored — likely in `data['boundary_id']` or similar.

**Step 3: Verify the mask is correct**

Print the number of aft-foil nodes being upweighted in the first batch to confirm the mask works:
```python
if step == 0 and cfg.aft_foil_loss_weight != 1.0:
    n_aft = aft_mask.sum().item()
    n_surf = surface_mask.sum().item()
    print(f"Aft-foil loss upweight: {n_aft}/{n_surf} surface nodes at {cfg.aft_foil_loss_weight}x")
```

### Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent nezuko --wandb_name "nezuko/aft-upweight-1.5-s42" \
  --wandb_group "round20/aft-foil-loss-upweight" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --aft_foil_loss_weight 1.5

# Seed 73
cd cfd_tandemfoil && python train.py \
  --agent nezuko --wandb_name "nezuko/aft-upweight-1.5-s73" \
  --wandb_group "round20/aft-foil-loss-upweight" \
  --seed 73 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --aft_foil_loss_weight 1.5
```

New flag: `--aft_foil_loss_weight 1.5`

### Report results

Table: p_in, p_oodc, p_tan, p_re for both seeds, 2-seed avg, baseline comparison, W&B run IDs.

**Focus on p_tan** — this is the target metric. Also watch p_in closely — if it degrades, the upweighting may be stealing gradient budget from fore-foil.

## Baseline

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in   | **11.979** | < 11.98 |
| p_oodc | **7.643**  | < 7.65  |
| p_tan  | **28.341** | < 28.34 |
| p_re   | **6.300**  | < 6.30  |

W&B: `hgml7i2r` (s42), `qic03vrg` (s73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent nezuko --wandb_name "nezuko/baseline" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Results

### Aft-Foil Surface Loss Upweighting: 1.5x

| Metric | Baseline | Seed 42 | Seed 73 | 2-seed Avg | Delta |
|--------|----------|---------|---------|------------|-------|
| p_in   | **11.979** | 11.9 | 12.6 | 12.25 | +0.27 ❌ |
| p_oodc | **7.643**  | 7.9  | 7.4  | 7.65  | +0.01 ❌ |
| p_tan  | **28.341** | 28.5 | 28.8 | 28.65 | +0.31 ❌ |
| p_re   | **6.300**  | 6.4  | 6.3  | 6.35  | +0.05 ❌ |

**W&B runs:**
- Seed 42: [`2i2u2sb4`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/2i2u2sb4)
- Seed 73: [`81who7b3`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/81who7b3)

**Commands used:**
```bash
# Seed 42
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent nezuko --wandb_name "nezuko/aft-upweight-1.5-s42" \
  --wandb_group "round20/aft-foil-loss-upweight" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --aft_foil_loss_weight 1.5

# Seed 73 — identical but --seed 73 --wandb_name "nezuko/aft-upweight-1.5-s73"
```

New flag: `--aft_foil_loss_weight 1.5`
Mask verification: 2924/8972 surface nodes (33%) identified as aft-foil and upweighted.
Best models at epochs 147 (s42) and 148 (s73).

### What happened

**Negative across all metrics.** The 1.5x aft-foil surface loss upweighting did not improve p_tan as hypothesized — it actually made it worse (+0.31).

- **p_tan worsened** (+0.31): The primary target metric degraded. Despite directing 1.5x more gradient to aft-foil nodes, the model did not produce better aft-foil predictions. This suggests the bottleneck for p_tan is not gradient allocation between fore/aft foils, but rather the model's ability to represent wake interactions.
- **p_in also degraded** (+0.27): Some gradient budget was shifted away from fore-foil, as expected. Seed 73 was particularly affected (12.6 vs 11.979 baseline).
- **p_oodc and p_re essentially flat** (within noise).

**Why it didn't work:** The surface loss already has tandem-specific boosting (`tandem_ramp` + `adaptive_boost`), which already upweights tandem samples. Adding aft-foil node-level weighting on top of the sample-level tandem boost may have created a compounding effect that over-emphasized aft-foil pressure at the expense of overall model quality. Additionally, the existing hard-node mining (epoch 30+) already targets the hardest surface nodes — many of which are likely aft-foil nodes — so the explicit upweighting is partially redundant.

### Suggested follow-ups

1. **Aft-foil upweighting only for non-tandem samples** — The tandem boost already handles tandem cases. Upweighting aft-foil nodes in non-tandem samples might provide complementary signal without the compounding issue.
2. **Lower weight (1.2x)** — 1.5x may be too aggressive given the existing tandem boost stack. A subtler nudge might help.
3. **Aft-foil-specific learning rate** — Instead of loss weighting, give the aft-foil SRF head a higher learning rate to let it specialize faster.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2253 because ADVISOR: **Closing — all 4 metrics regress on 2-seed average.**

| Metric | Baseline | 2-seed Avg | Δ |
|--------|----------|-----------|---|
| p_in | 11.979 | 12.247 | +2.2% ✗ |
| p_oodc | 7.643 | 7.647 | +0.05% ✗ |
| p_tan | 28.341 | 28.646 | +1.1% ✗ |
| p_re | 6.300 | 6.324 | +0.4% ✗ |

W&B runs: 2i2u2sb4 (s42), 81who7b3 (s73) — verified.

**Analysis:** The 1.5x aft-foil loss weight compounds with the existing tandem_ramp + adaptive_boost + PCGrad 3-way gradient mechanisms. The baseline already aggressively upweights tandem/hard cases at the sample level — adding node-level upweighting on top creates gradient instability. The high p_in seed variance (11.87 vs 12.62 — a 0.75 gap) confirms the optimizer is sensitive to this compounding. Most critically, p_tan (the target metric) WORSENED by +0.3, suggesting the bottleneck is representational capacity for wake interactions, not gradient allocation to aft-foil nodes.

**Key insight:** The existing multi-objective gradient pipeline (PCGrad extreme + tandem ramp) has saturated the gradient budget allocated to hard samples. Further loss weighting is redundant. Future aft-foil improvements should come from giving the model new information or structural capacity, not more gradient pressure.

Good analysis on the interaction with tandem_ramp, nezuko. Moving you to a fresh experiment..
